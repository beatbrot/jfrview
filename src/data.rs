use jfrs::reader::{
    event::{Accessor, Event},
    value_descriptor::ValueDescriptor,
};

#[derive(Debug)]
#[allow(dead_code)]
pub struct ExecutionSample {
    pub start_time: i64,
    pub thread: Thread,
    pub stack_trace: StackTrace,
}

impl From<Event<'_>> for ExecutionSample {
    fn from(value: Event) -> Self {
        let v = value.value();
        let start_time: i64 = extract_primitive(&v, "startTime");
        let thread: Thread = value.value().get_field("sampledThread").unwrap().into();
        let stack_trace: StackTrace = value.value().get_field("stackTrace").unwrap().into();
        return Self {
            start_time,
            thread,
            stack_trace,
        };
    }
}

#[derive(Debug)]
#[allow(dead_code)]
pub struct Thread {
    pub java_name: String,
    pub java_thread_id: i64,
}

impl From<Accessor<'_>> for Thread {
    fn from(value: Accessor<'_>) -> Self {
        let java_name = extract_primitive::<&str>(&value, "javaName").to_string();
        let java_thread_id: i64 = extract_primitive(&value, "javaThreadId");
        return Thread {
            java_name,
            java_thread_id,
        };
    }
}

impl PartialEq for Thread {
    fn eq(&self, other: &Self) -> bool {
        self.java_thread_id == other.java_thread_id
    }
}

#[derive(Debug)]
#[allow(dead_code)]
pub struct StackTrace {
    pub truncated: bool,
    pub frames: Vec<StackFrame>,
}

impl From<Accessor<'_>> for StackTrace {
    fn from(value: Accessor<'_>) -> Self {
        let truncated: bool = extract_primitive(&value, "truncated");
        let mut frames: Vec<StackFrame> = value
            .get_field("frames")
            .unwrap()
            .as_iter()
            .unwrap()
            .map(|a| a.into())
            .collect();
        frames.reverse();

        return Self { truncated, frames };
    }
}

#[allow(dead_code)]
pub struct StackFrame {
    pub line_number: i32,
    pub bytecode_index: i32,
    pub method: Method,
}

impl From<Accessor<'_>> for StackFrame {
    fn from(value: Accessor<'_>) -> Self {
        let bytecode_index: i32 = extract_primitive(&value, "bytecodeIndex");
        let line_number: i32 = extract_primitive(&value, "lineNumber");
        let method: Method = value.get_field("method").map(|v| v.into()).unwrap();
        return Self {
            method,
            line_number,
            bytecode_index,
        };
    }
}

impl std::fmt::Debug for StackFrame {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}:{}", self.method, self.line_number)
    }
}

#[allow(dead_code)]
#[derive(Hash, PartialEq, Eq, Clone)]
pub struct Method {
    name: String,
    class: Class,
}

impl std::fmt::Debug for Method {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:{}", self.class.short_name(), self.name)
    }
}

impl From<Accessor<'_>> for Method {
    fn from(value: Accessor<'_>) -> Self {
        let name = extract_symbol(&value, "name");
        let class: Class = value.get_field("type").map(|v| v.into()).unwrap();
        return Self { name, class };
    }
}

#[derive(Debug, Hash, PartialEq, Eq, Clone)]
#[allow(dead_code)]
pub struct Class {
    name: String,
}

impl Class {
    pub fn short_name(&self) -> String {
        let segments: Vec<&str> = self.name.split('/').collect();
        let mut result = String::new();

        for (i, segment) in segments.iter().enumerate() {
            if i != 0 {
                result.push('/');
            }

            if i == segments.len() - 1 {
                result.push_str(segment);
            } else if !segment.is_empty() {
                result.push(segment.chars().next().unwrap());
            }
        }

        return result;
    }
}

impl std::fmt::Display for Class {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}

impl From<Accessor<'_>> for Class {
    fn from(value: Accessor<'_>) -> Self {
        let name = extract_symbol(&value, "name");
        return Self { name };
    }
}

fn extract_symbol(value: &Accessor, name: &str) -> String {
    let str: &str = extract_primitive(&value.get_field(name).unwrap(), "string");
    return str.to_string();
}

fn extract_primitive<'a, T>(value: &Accessor<'a>, name: &str) -> T
where
    T: TryFrom<&'a ValueDescriptor>,
{
    return value
        .get_field(name)
        .and_then(|v| <T>::try_from(v.value).ok())
        .unwrap();
}
