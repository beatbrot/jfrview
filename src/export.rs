use std::{
    hash::Hash,
    io::{Read, Seek},
};

use crate::data::ExecutionSample;
use indexmap::{set::MutableValues, IndexSet};
use serde::Serialize;

#[derive(Serialize)]
pub struct Sample {
    name: String,
    value: usize,
    children: IndexSet<Sample>,
}

impl PartialEq for Sample {
    fn eq(&self, other: &Self) -> bool {
        return self.name.eq(&other.name);
    }
}

impl Eq for Sample {}

impl Hash for Sample {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.name.hash(state);
    }
}

impl Sample {
    pub fn from_file<T>(input: T, include_native: bool, threads: bool) -> anyhow::Result<Self>
    where
        T: Read + Seek,
    {
        let mut result = Self::new("root".to_string());
        ExecutionSample::visit_events(input, |e| result.add_sample(&e, include_native, threads))?;

        Ok(result)
    }

    fn add_sample(&mut self, sample: &ExecutionSample, include_native: bool, threads: bool) {
        if sample.native && !include_native {
            return;
        }
        let mut curr = self;
        curr.value += 1;
        if threads {
            curr = curr.thread_root(sample);
            curr.value += 1;
        }
        for frame in sample.stack_trace.frames.iter() {
            let name = format!("{:?}", frame.method);
            let (idx, _) = curr.children.insert_full(Self::new(name));
            let sample = curr.children.get_index_mut2(idx).unwrap();
            sample.value += 1;
            curr = sample;
        }
    }

    fn thread_root(&mut self, sample: &ExecutionSample) -> &mut Sample {
        let thread_name = sample
            .thread
            .java_name
            .clone()
            .unwrap_or_else(|| sample.thread.java_thread_id.to_string());
        let (idx, _) = self.children.insert_full(Self::new(thread_name));
        return self.children.get_index_mut2(idx).unwrap();
    }

    fn new(name: String) -> Self {
        Self {
            name,
            value: 0,
            children: Default::default(),
        }
    }
}

#[cfg(test)]
impl core::fmt::Display for Sample {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        fn printer(
            f: &mut std::fmt::Formatter<'_>,
            indent: usize,
            sample: &Sample,
        ) -> std::fmt::Result {
            writeln!(
                f,
                "{}{}: ({})",
                "| ".repeat(indent),
                sample.name,
                sample.value
            )?;
            for c in &sample.children {
                printer(f, indent + 1, c)?;
            }
            Ok(())
        }
        for c in &self.children {
            printer(f, 0, c)?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use insta::{assert_snapshot, glob};
    use std::fs::File;

    use crate::export::Sample;

    #[test]
    fn parse_without_panic_nonnative() {
        glob!("../test-data", "*.jfr", |path| {
            let f = File::open(path).unwrap();
            let flame_graph = Sample::from_file(f, false);
            let file_name = path.file_name().unwrap().to_string_lossy();
            insta::with_settings!({
                omit_expression => true,
                description => format!("non-native, {}", file_name)
            }, {
                assert_snapshot!(flame_graph.unwrap());
            });
        });
    }

    #[test]
    fn parse_without_panic_native() {
        glob!("../test-data", "*.jfr", |path| {
            let f = File::open(path).unwrap();
            let flame_graph = Sample::from_file(f, true);
            let file_name = path.file_name().unwrap().to_string_lossy();
            insta::with_settings!({
                omit_expression => true,
                description => format!("native, {}", file_name)
            }, {
                assert_snapshot!(flame_graph.unwrap());
            });
        });
    }

    #[test]
    fn test_invalid() -> anyhow::Result<()> {
        let file = File::open("test-data/invalid.jfr.fail")?;
        assert!(Sample::from_file(file, false).is_err());
        Ok(())
    }
}
