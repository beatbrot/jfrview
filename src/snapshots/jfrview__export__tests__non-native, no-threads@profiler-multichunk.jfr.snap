---
source: src/export.rs
input_file: test-data/profiler-multichunk.jfr
---
:/usr/lib/aarch64-linux-gnu/libc.so.6: (5706)
| :/usr/lib/aarch64-linux-gnu/libc.so.6: (5706)
| | libjvm.so:thread_native_entry: (5706)
| | | libjvm.so:Thread::call_run: (5706)
| | | | libjvm.so:ConcurrentGCThread::run: (900)
| | | | | libjvm.so:G1ConcurrentMarkThread::run_service: (300)
| | | | | | libjvm.so:Monitor::wait: (300)
| | | | | | | libjvm.so:Monitor::IWait: (300)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (300)
| | | | | | | | | libc.so.6:pthread_cond_wait: (300)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
| | | | | libjvm.so:G1ConcurrentRefineThread::run_service: (300)
| | | | | | libjvm.so:Monitor::wait: (300)
| | | | | | | libjvm.so:Monitor::IWait: (300)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (300)
| | | | | | | | | libc.so.6:pthread_cond_wait: (300)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
| | | | | libjvm.so:G1YoungRemSetSamplingThread::run_service: (300)
| | | | | | libjvm.so:Monitor::wait: (300)
| | | | | | | libjvm.so:Monitor::IWait: (300)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (300)
| | | | | | | | | libc.so.6:pthread_cond_timedwait: (300)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
| | | | libjvm.so:JavaThread::thread_main_inner: (2405)
| | | | | libjvm.so:ServiceThread::service_thread_entry: (301)
| | | | | | libjvm.so:Monitor::wait: (301)
| | | | | | | libjvm.so:Monitor::IWait: (301)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (301)
| | | | | | | | | libc.so.6:pthread_cond_wait: (301)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301)
| | | | | libjvm.so:CompileBroker::compiler_thread_loop: (602)
| | | | | | libjvm.so:CompileQueue::get: (602)
| | | | | | | libjvm.so:Monitor::wait: (602)
| | | | | | | | libjvm.so:Monitor::IWait: (602)
| | | | | | | | | libjvm.so:os::PlatformEvent::park: (602)
| | | | | | | | | | libc.so.6:pthread_cond_timedwait: (602)
| | | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (602)
| | | | | libjvm.so:attach_listener_thread_entry: (300)
| | | | | | libjvm.so:AttachListener::dequeue: (300)
| | | | | | | libjvm.so:LinuxAttachListener::dequeue: (300)
| | | | | | | | libc.so.6:accept: (300)
| | | | | libjvm.so:signal_thread_entry: (301)
| | | | | | libjvm.so:os::signal_wait: (301)
| | | | | | | libjvm.so:PosixSemaphore::wait: (301)
| | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301)
| | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301)
| | | | | libjvm.so:recorderthread_entry: (300)
| | | | | | libjvm.so:Monitor::wait: (300)
| | | | | | | libjvm.so:Monitor::IWait: (300)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (300)
| | | | | | | | | libc.so.6:pthread_cond_wait: (300)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
| | | | | libjvm.so:JvmtiAgentThread::call_start_function: (300)
| | | | | | libasyncProfiler.so:Profiler::timerLoop: (300)
| | | | | | | libc.so.6:pthread_cond_timedwait: (300)
| | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
| | | | | libjvm.so:NMethodSweeper::sweeper_loop: (301)
| | | | | | libjvm.so:Monitor::wait: (301)
| | | | | | | libjvm.so:Monitor::IWait: (301)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (301)
| | | | | | | | | libc.so.6:pthread_cond_timedwait: (301)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301)
| | | | libjvm.so:GangWorker::loop: (1500)
| | | | | libjvm.so:SemaphoreGangTaskDispatcher::worker_wait_for_task: (1500)
| | | | | | libjvm.so:PosixSemaphore::wait: (1500)
| | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (1500)
| | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (1500)
| | | | libjvm.so:VMThread::run: (300)
| | | | | libjvm.so:VMThread::loop: (300)
| | | | | | libjvm.so:Monitor::wait: (300)
| | | | | | | libjvm.so:Monitor::IWait: (300)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (300)
| | | | | | | | | libc.so.6:pthread_cond_timedwait: (300)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
| | | | libjvm.so:WatcherThread::run: (301)
| | | | | libjvm.so:WatcherThread::sleep: (301)
| | | | | | libjvm.so:Monitor::wait: (301)
| | | | | | | libjvm.so:Monitor::IWait: (301)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (301)
| | | | | | | | | libc.so.6:pthread_cond_timedwait: (301)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301)
| | | | libjvm.so:JfrThreadSampler::run: (300)
| | | | | libjvm.so:PosixSemaphore::wait: (300)
| | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
| | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
java.lang.Thread:run: (1379)
| java.util.concurrent.ThreadPoolExecutor$Worker:run: (1079)
| | java.util.concurrent.ThreadPoolExecutor:runWorker: (1079)
| | | Example$$Lambda$5.575335780:run: (599)
| | | | Example:lambda$main$3: (599)
| | | | | Example:sleep: (599)
| | | | | | java.lang.Thread:yield: (599)
| | | | | | | libc.so.6:__sched_yield: (598)
| | | | | | | libjvm.so:sched_yield@plt: (1)
| | | Example$$Lambda$4.1500056228:run: (480)
| | | | Example:lambda$main$2: (480)
| | | | | Example$Writer:println: (60)
| | | | | | sun.nio.ch.FileChannelImpl:force: (60)
| | | | | | | sun.nio.ch.FileDispatcherImpl:force: (60)
| | | | | | | | sun.nio.ch.FileDispatcherImpl:force0: (60)
| | | | | | | | | libnio.so:Java_sun_nio_ch_FileDispatcherImpl_force0: (60)
| | | | | | | | | | libc.so.6:fsync: (60)
| | | | | java.util.concurrent.LinkedBlockingQueue:poll: (420)
| | | | | | java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject:awaitNanos: (420)
| | | | | | | java.util.concurrent.locks.LockSupport:parkNanos: (420)
| | | | | | | | jdk.internal.misc.Unsafe:park: (420)
| | | | | | | | | libjvm.so:Unsafe_Park: (420)
| | | | | | | | | | libjvm.so:Parker::park: (420)
| | | | | | | | | | | libc.so.6:pthread_cond_timedwait: (420)
| | | | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (420)
| jdk.jfr.internal.PlatformRecorder$$Lambda$55.205465701:run: (300)
| | jdk.jfr.internal.PlatformRecorder:lambda$startDiskMonitor$1: (300)
| | | jdk.jfr.internal.PlatformRecorder:periodicTask: (300)
| | | | jdk.jfr.internal.PlatformRecorder:takeNap: (300)
| | | | | java.lang.Object:wait: (300)
| | | | | | libjvm.so:JVM_MonitorWait: (300)
| | | | | | | libjvm.so:ObjectSynchronizer::wait: (300)
| | | | | | | | libjvm.so:ObjectMonitor::wait: (300)
| | | | | | | | | libjvm.so:os::PlatformEvent::park: (300)
| | | | | | | | | | libc.so.6:pthread_cond_timedwait: (299)
| | | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (299)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (1)
:unknown_Java: (1)
| libc.so.6:clock_gettime: (1)
java.lang.ref.Finalizer$FinalizerThread:run: (301)
| java.lang.ref.ReferenceQueue:remove: (301)
| | java.lang.ref.ReferenceQueue:remove: (301)
| | | java.lang.Object:wait: (301)
| | | | libjvm.so:JVM_MonitorWait: (301)
| | | | | libjvm.so:ObjectSynchronizer::wait: (301)
| | | | | | libjvm.so:ObjectMonitor::wait: (301)
| | | | | | | libjvm.so:os::PlatformEvent::park: (301)
| | | | | | | | libc.so.6:pthread_cond_wait: (301)
| | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301)
Example:main: (300)
| java.util.concurrent.ThreadPoolExecutor:awaitTermination: (300)
| | java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject:awaitNanos: (300)
| | | java.util.concurrent.locks.LockSupport:parkNanos: (300)
| | | | jdk.internal.misc.Unsafe:park: (300)
| | | | | libjvm.so:Unsafe_Park: (300)
| | | | | | libjvm.so:Parker::park: (300)
| | | | | | | libc.so.6:pthread_cond_timedwait: (300)
| | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
jdk.internal.misc.InnocuousThread:run: (300)
| java.lang.Thread:run: (300)
| | jdk.internal.ref.CleanerImpl:run: (300)
| | | java.lang.ref.ReferenceQueue:remove: (300)
| | | | java.lang.Object:wait: (300)
| | | | | libjvm.so:JVM_MonitorWait: (300)
| | | | | | libjvm.so:ObjectSynchronizer::wait: (300)
| | | | | | | libjvm.so:ObjectMonitor::wait: (300)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (300)
| | | | | | | | | libc.so.6:pthread_cond_timedwait: (300)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
:unknown: (300)
| java:_start: (300)
| | libc.so.6:__libc_start_main: (300)
| | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
| | | | java:main: (300)
| | | | | libjli.so:JLI_Launch: (300)
| | | | | | libjli.so:ContinueInNewThread: (300)
| | | | | | | libjli.so:CallJavaMainInNewThread: (300)
| | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
| | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
java.lang.ref.Reference$ReferenceHandler:run: (301)
| java.lang.ref.Reference:processPendingReferences: (301)
| | java.lang.ref.Reference:waitForReferencePendingList: (301)
| | | libjvm.so:JVM_WaitForReferencePendingList: (301)
| | | | libjvm.so:Monitor::wait: (301)
| | | | | libjvm.so:Monitor::IWait: (301)
| | | | | | libjvm.so:os::PlatformEvent::park: (301)
| | | | | | | libc.so.6:pthread_cond_wait: (301)
| | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301)
java.util.TimerThread:run: (300)
| java.util.TimerThread:mainLoop: (300)
| | java.lang.Object:wait: (300)
| | | java.lang.Object:wait: (300)
| | | | libjvm.so:JVM_MonitorWait: (300)
| | | | | libjvm.so:ObjectSynchronizer::wait: (300)
| | | | | | libjvm.so:ObjectMonitor::wait: (300)
| | | | | | | libjvm.so:os::PlatformEvent::park: (300)
| | | | | | | | libc.so.6:pthread_cond_wait: (300)
| | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300)
